### PR TITLE
Integrate DSViz into Keras WandbCallback

### PR DESCRIPTION
**Pull request by tssweeney**
Description
-----------

The goal of this PR is to lower the barrier of entry for users to start using DSViz by integrating with the Keras Callback. We anticipate that this is mostly to get people started and as the complexity of their project increases, we expect users to begin using the Table/logging apis directly. In any case, this PR adds the following functionality (in the form of constructor paramters) to the `WandbCallback`:

- `log_validation` (a boolean flag) has been repurposed. It is currently busted. Now, if True, we will automatically create the following tables:
   - `validation_data` is a table created with columns that map to the validation data used in the `model.fit` function. Depending on the other parameters provided, this Table may not be created if an existing table of validation data already exists. More on this later. The columns of this table are both dependent on the shape of the validation data as well as the provided processor functions. In the base case, it will contain one column called 'input' and one column called 'target' - assuming no processor inference, no provided processors, and the x/y data are simply list-like series of examples. If the validation inputs and/or targets are dictionaries, then the keys of the dictionaries will be used as column names. Furthermore, if processors are provided, or inferred, additional derivative columns will be included. The `validation_data` table is logged to its own artifact.
   - `validation_predictions` is a table created with one column that links back to the `validation_data` table, and at least one additional column for the output of the model, given the respective inputs. Again, the columns are dependent on other values. If the output of the model is a keyed dictionary, then the keys will be used as column names. If the user provides a prediction processor, or one is inferred, then additional columns will be provided. `validation_predictions` tables are logged directly to the run itself, once per epoch. This creates an underlying artifact per epoch and allows the user to explore the table using the Weave/DSViz Table Visualizer. 
- `validation_indexes`: An optional list of `wandb.data_types._TableIndex` which map validation data rows to an already logged Table. This is useful in the case that the user has a dataset already logged and just wants to map model predictions to this table, rather than logging the entire validation set every run.
- `validation_row_processor`: A function to apply to the validation data, commonly used to visualize the data. The function will receive an ndx (int) and a row (dict). If your model has a single input, then row["input"] will be the input data for the row. Else, it will be keyed based on the name of the input slot. If your fit function takes a single target, then row["target"] will be the target data for the row. Else, it will be keyed based on the name of the output slots. For example, if your input data is a single ndarray, but you wish to visualize the data as an Image, then you can provide `lambda ndx, row: {"img": wandb.Image(row["input"])}` as the processor.
- `output_row_processor`: same as validation_row_processor, but applied to the model's output. `row["output"]` will contain the results of the model output.
- `infer_missing_processors`: Determines if validation_row_processor and output_row_processor should be inferred if missing. Defaults to True. If `labels` are provided, we will attempt to infer classification-type processors where appropriate.

Moreover, all of this functionality is exposed via a new class `ValidationDataLogger` internally which can be used in future library integrations to make the integration seamless and similar. See how thin the actual `WandbCallback` modifications are.

Now, a few examples (assume this is a notebook)

Let's say you run the following script:

### Setup (Unrelated to this change)
```
import os
import random

import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.datasets import cifar10
from sklearn.model_selection import train_test_split

import wandb
from wandb.keras import WandbCallback

project = "keras-integration_v3.15"

default_config = {  # and include hyperparameters and metadata
        "learning_rate": 0.005,
        "epochs": 3,
        "batch_size": 64,
        "loss_function": "sparse_categorical_crossentropy",
        "architecture": "CNN",
        "dataset": "CIFAR-10",
        "train_count": 500,
        "test_count": 10,
    }

CLASS_NAMES = [
    "airplane",
    "automobile",
    "bird",
    "cat",
    "deer",
    "dog",
    "frog",
    "horse",
    "ship",
    "truck",
]


def Model():
    inputs = keras.layers.Input(shape=(32, 32, 3))

    x = keras.layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(inputs)
    x = keras.layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)
    x = keras.layers.MaxPooling2D(pool_size=2)(x)

    x = keras.layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)
    x = keras.layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)

    x = keras.layers.GlobalAveragePooling2D()(x)

    x = keras.layers.Dense(128, activation="relu")(x)
    x = keras.layers.Dense(32, activation="relu")(x)

    outputs = keras.layers.Dense(len(CLASS_NAMES), activation="softmax")(x)

    return keras.models.Model(inputs=inputs, outputs=outputs)
```

### Base Case
In this "cell" we will run 2 Runs which simply set `log_evaluation` to True. The reason 2 are run is to show how deduplication works. The only "special" line in this cell is `log_evaluation=True` - everything else is just normal model training script.
- Run Result 1: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/brl50x7d
- Run Result 2: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/ejymkoqj

```
for _ in range(2):
    with wandb.init(
        project=project,
        config=default_config,
    ) as run:
        config = wandb.config
        (x_train, y_train), (x_test, y_test) = cifar10.load_data()
        x_train, x_test = x_train[::5] / 255.0, x_test / 255.0
        y_train, y_test = y_train[::5], y_test[::5]
        (x_train, y_train) = (x_train[:config.train_count], y_train[:config.train_count])
        (x_test, y_test) = (x_test[:config.test_count], y_train[:config.test_count])

        model = Model()
        optimizer = tf.keras.optimizers.Adam(config.learning_rate)
        model.compile(optimizer, config.loss_function, metrics=["acc"])

        _ = model.fit(
            x_train,
            y_train,
            epochs=config.epochs,
            batch_size=config.batch_size,
            validation_data=(x_test, y_test),
            callbacks=[WandbCallback(
                log_evaluation=True,
            )],
        )

        run.finish()
```

| Resulting Artifact Graph |
|---|
| Screen Shot 2021-04-06 at 2 07 37 PM  | 

|   | Table Viz  | Table Schema  |
|---|---|---|
| Validation Data | Screen Shot 2021-04-06 at 2 11 25 PM | Screen Shot 2021-04-06 at 2 11 51 PM |
| Validation Predictions | Screen Shot 2021-04-06 at 2 13 13 PM | Screen Shot 2021-04-06 at 2 13 30 PM |

### With Class Labels
In this "cell" we will run 2 Runs which also set "labels" to an array of class labels. The inferred processors will use this to build class tables and links
- Run Result 1: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/1zi3smyc
- Run Result 2: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/2wv33ptq

```
for _ in range(2):
    with wandb.init(
        project=project,
        config=default_config,
    ) as run:
        config = wandb.config
        (x_train, y_train), (x_test, y_test) = cifar10.load_data()
        x_train, x_test = x_train[::5] / 255.0, x_test / 255.0
        y_train, y_test = y_train[::5], y_test[::5]
        (x_train, y_train) = (x_train[:config.train_count], y_train[:config.train_count])
        (x_test, y_test) = (x_test[:config.test_count], y_train[:config.test_count])

        model = Model()
        optimizer = tf.keras.optimizers.Adam(config.learning_rate)
        model.compile(optimizer, config.loss_function, metrics=["acc"])

        _ = model.fit(
            x_train,
            y_train,
            epochs=config.epochs,
            batch_size=config.batch_size,
            validation_data=(x_test, y_test),
            callbacks=[WandbCallback(
                labels=CLASS_NAMES,
                log_evaluation=True,
            )],
        )

        run.finish()
```

| Resulting Artifact Graph |
|---|
| Screen Shot 2021-04-06 at 2 26 33 PM | 

|   | Table Viz  | Table Schema  |
|---|---|---|
| Validation Data | Screen Shot 2021-04-06 at 2 27 08 PM | Screen Shot 2021-04-06 at 2 27 22 PM |
| Validation Predictions | Screen Shot 2021-04-06 at 2 28 30 PM | Screen Shot 2021-04-06 at 2 28 56 PM |

### With Processors
In this "cell" we will run 2 Runs which specify user-provided processors
- Run Result 1: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/1oqsxok3
- Run Result 2: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/338kegjs

```
for _ in range(2):
    with wandb.init(
        project=project,
        config=default_config,
    ) as run:
        config = wandb.config
        (x_train, y_train), (x_test, y_test) = cifar10.load_data()
        x_train, x_test = x_train[::5] / 255.0, x_test / 255.0
        y_train, y_test = y_train[::5], y_test[::5]
        (x_train, y_train) = (x_train[:config.train_count], y_train[:config.train_count])
        (x_test, y_test) = (x_test[:config.test_count], y_train[:config.test_count])

        model = Model()
        optimizer = tf.keras.optimizers.Adam(config.learning_rate)
        model.compile(optimizer, config.loss_function, metrics=["acc"])
        class_table = wandb.Table(columns=[], data=[])
        class_table.add_column("name", CLASS_NAMES)

        _ = model.fit(
            x_train,
            y_train,
            epochs=config.epochs,
            batch_size=config.batch_size,
            validation_data=(x_test, y_test),
            callbacks=[WandbCallback(
                log_evaluation=True,
                validation_row_processor=lambda ndx, row: {
                    "input:image": wandb.Image(row["input"]),
                    "target:class": class_table.index_ref(row["target"])
                },
                prediction_row_processor=lambda ndx, row: {
                    "output:class": class_table.index_ref(np.argmax(row["output"])), 
                    "output:logits": {class_name: value for (class_name, value) in zip(CLASS_NAMES, row["output"].tolist())}
                }
            )],
        )

        run.finish()
```

These results are very similar to the prior cell's

### Create a Dataset
In this cell we simply create a dataset for use in the next cell
- Run Page: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/117459n7

```
with wandb.init(project=project) as run:
    (x_train, y_train), _ = cifar10.load_data()
    x_train = x_train[::5] / 255.0
    y_train = y_train[::5]
    p = np.random.permutation(len(x_train))
    (x_train, y_train) = x_train[p], y_train[p]
    (x_train, y_train) = (x_train[:5000], y_train[:5000])

    # Create the dataset
    ds = wandb.Artifact("cifar_dataset", "dataset")

    # Add the class table
    class_table = wandb.Table(columns=[], data=[])
    class_table.add_column("class_name", CLASS_NAMES)
    ds["class_table"] = class_table

    # Add the train table
    train_table = wandb.Table(columns=[], data=[])
    train_table.add_column("x_train", x_train)
    train_table.add_column("y_train", y_train)
    train_table.add_computed_columns(lambda ndx, row:{
        "x_train_img": wandb.Image(row["x_train"]),
        "y_train_cls": class_table.index_ref(row["y_train"])
    })
    ds["train_table"] = train_table

    # Save the dataset
    ds.save()
```

### Using Index Links
In this "cell" we will run 2 Runs which specify `validation_indexes` based on the dataset created above
- Run Result 1: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/2lny9ueo
- Run Result 2: https://wandb.ai/timssweeney/keras-integration_v3.15/runs/18k4w4py

```
for _ in range(2):
    with wandb.init(
        project=project,
        config=default_config,
    ) as run:
        config = wandb.config
        art = run.use_artifact("cifar_dataset:latest")
        train_table = art.get("train_table")
        class_table = art.get("class_table")
        x = train_table.get_column("x_train", convert_to="numpy")
        y = train_table.get_column("y_train", convert_to="numpy")
        ids = train_table.get_index()
        x_train, x_test, y_train, y_test, id_train, id_test = train_test_split(x, y, ids, test_size=0.33)


        model = Model()
        optimizer = tf.keras.optimizers.Adam(config.learning_rate)
        model.compile(optimizer, config.loss_function, metrics=["acc"])

        _ = model.fit(
            x_train,
            y_train,
            epochs=config.epochs,
            batch_size=config.batch_size,
            validation_data=(x_test, y_test),
            callbacks=[WandbCallback(
                log_evaluation=True,
                labels=class_table.get_column("class_name"),
                validation_indexes=id_test,
            )],
        )

        run.finish()
```

Here, the tables are nearly the same, but the cool part is the graph. Despite shuffling, we maintain a single dataset of truth.

Screen Shot 2021-04-06 at 2 31 36 PM

Testing
-------

How was this PR tested? Notebook usage and a giant unit test suite

### CODE REVIEW

**Comment by charlesfrye** [comment]
FYI the data setup code for the examples has a small bug: the `y_test` subset is pulled from `y_train` instead of `y_test`. Notice that the val images labeled "frog"/"automobile" include things that are decidedly not frogs/autos.

image

I was only able to notice this bug because of DSViz!

The correct code is:
```
for _ in range(2):
    with wandb.init(
        project=project,
        config=default_config,
    ) as run:
        config = wandb.config
        (x_train, y_train), (x_test, y_test) = cifar10.load_data()
        x_train, x_test = x_train[::5] / 255.0, x_test / 255.0
        y_train, y_test = y_train[::5], y_test[::5]
        (x_train, y_train) = (x_train[:config.train_count], y_train[:config.train_count])
        (x_test, y_test) = (x_test[:config.test_count], y_test[:config.test_count])
```

File: wandb/sdk_py27/wandb_artifacts.py [hunk 1]
```diff
@@ -430,7 +432,12 @@ def add_reference(
 
         return manifest_entries
 
-    def add(self, obj, name):
+    def add(
+        self,
+        obj,
+        name,
+        _rename_deterministically = False,
```
Comment (shawnlewis) [inline wandb/sdk_py27/wandb_artifacts.py:439]: Is there a way to avoid leaking this "_rename_deterministically" thing to the user?

File: wandb/sdk_py27/wandb_artifacts.py [hunk 2]
```diff
@@ -472,7 +479,11 @@ def add(self, obj, name):
         entry = self._manifest.get_entry_by_path(name)
         if entry is not None:
             return entry
-        with self.new_file(name) as f:
+        file_path = os.path.join(ARTIFACT_TMP.name, str(id(self)), name)
```
Comment (shawnlewis) [inline wandb/sdk_py27/wandb_artifacts.py:482]: Does this mean we're going to write a bunch of files to wherever the temp directory happens to be and leave them there?

I believe maybe we're also doing this with the fixes Anni put in for the cache race.

Just be aware of the potential to blow up wherever the temp file system is with too much data. User's will probably want control of where this happens if there's any risk of that.

File: wandb/integration/keras/keras.py [hunk 3]
```diff
@@ -500,11 +544,22 @@ def on_test_batch_end(self, batch, logs=None):
         pass
 
     def on_train_begin(self, logs=None):
-        pass
+        if self.log_evaluation:
+            try:
+                self.validation_data_logger = ValidationDataLogger(
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply wandb/integration/keras/keras.py:549]: Thank you!

File: wandb/sdk/integration_utils/data_logging.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        self.local_validation_artifact = local_validation_artifact
+        self.input_col_name = input_col_name
+
+    def make_predictions(
+        self, predict_fn: Callable
+    ) -> Union[Sequence, Dict[str, Sequence]]:
+        """Produces predictions by passing `validation_inputs` to `predict_fn`.
+
+        Args:
+            predict_fn (Callable): Any function which can accept `validation_inputs` and product
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply wandb/sdk/integration_utils/data_logging.py:153]: Wow, nice catch. Fixed in next commit

File: wandb/sdk_py27/wandb_artifacts.py (hunk 1 above)
Comment (tssweeney) [reply wandb/sdk_py27/wandb_artifacts.py]: removed!

File: wandb/sdk_py27/wandb_artifacts.py (hunk 2 above)
Comment (tssweeney) [reply wandb/sdk_py27/wandb_artifacts.py]: added explicit cleanup

**Comment by tssweeney** [comment]
Running regression now

**Comment by tssweeney** [comment]
Hold off on merging - found one potential issue in integration test
